### ISSUE TITLE
Tracking Issue for `Thread::{into_raw, from_raw}`

### ISSUE THREAD

**Issue by ibraheemdev** [rust-lang/rust#97523]
Feature gate: `#![feature(thread_raw)]`

This is a tracking issue for `Thread::into_raw` and `Thread::from_raw`.

These functions allow conversions between a `Thread` and a raw pointer, which is useful in concurrent code as it allows the use of `Thread` with `AtomicPtr`.

### Public API

```rust
// std::thread

impl Thread {
    fn into_raw(self) -> *const ();
    unsafe fn from_raw(ptr: *const ()) -> Thread;    
}
```

### Steps / History

- [ ] Implementation: #...
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by yanchith** [comment]
Not sure If I am missing something, but `into_raw` requires the thread by value, and that is not always available. E.g. `JoinHandle::thread` only gives out a shared reference.

How can I unpark a thread for which I just have a JoinHandle?

**Comment by ibraheemdev** [comment]
@yanchith `Thread` is cloneable.

**Comment by yanchith** [comment]
Yeah, thanks! I didn't notice Thread was Clone. I am perfectly happy with that now.

`Thread::as_raw`: sure, if it can be designed such that the data inside the Arc inside Thread doesn't get dropped meanwhile. `Thread::into_inner` leaks the Arc, so that's taken care of.

**Comment by zesterer** [comment]
As a point of interest: is it legal to use `Thread::from_raw` if the pointer is shared in a multi-threaded context? A lot of code assumes this to be true for `Arc::from_raw`, and of course the implementation uses an `Arc`, but it is non-obvious.

Use of `Thread::from_raw` in a multi-threaded context might result in two `Thread`s existing that both point to the same block of memory, but with the strong count still being 1 (even if momentarily).

An API like `fn Thread::from_raw_ref(ptr: *const ()) -> &Thread` would make it much easier to use this API properly, although the same is probably true of `Arc`.
